processor/cloningfanoutconnector.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processor
+
+import (
+	"context"
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/proto"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+)
+
+// This file contains implementations of cloning Trace/Metrics connectors
+// that fan out the data to multiple other consumers. Cloning connectors create
+// clones of data before fanning out, which ensures each consumer gets their
+// own copy of data and is free to modify it.
+
+// NewMetricsCloningFanOutConnector wraps multiple metrics consumers in a single one.
+func NewMetricsCloningFanOutConnector(mcs []consumer.MetricsConsumer) consumer.MetricsConsumer {
+	return metricsCloningFanOutConnector(mcs)
+}
+
+type metricsCloningFanOutConnector []consumer.MetricsConsumer
+
+var _ consumer.MetricsConsumer = (*metricsCloningFanOutConnector)(nil)
+
+// ConsumeMetricsData exports the MetricsData to all consumers wrapped by the current one.
+func (mfc metricsCloningFanOutConnector) ConsumeMetricsData(ctx context.Context, md consumerdata.MetricsData) error {
+	var errs []error
+
+	// Fan out to first len-1 consumers.
+	for i := 0; i < len(mfc)-1; i++ {
+		// Create a clone of data. We need to clone because consumers may modify the data.
```

Comment:
I considered that and rejected. It would mean we cannot keep natural processing order because the original data needs to be passed to the last (timeline-wise) consumer, because that processor may modify the original data.
To ensure we timeline wise we call consumers in the order they are listed in `mfc` I had to give the original to the last one (index-wise).

Reply:
makes sense